runtime/storage/cache/cache.go

```diff
@@ -0,0 +1,267 @@
+	VolatileLFU EvictionPolicy = "volatile-lfu"
+
+	// VolatileTTL removes keys with an expiration set and the shortest time to live (TTL).
+	// It behaves like NoEviction if there are no keys to evict with an expiration set.
+	VolatileTTL EvictionPolicy = "volatile-ttl"
+
+	// VolatileRandom randomly removes keys with an expiration set.
+	// It behaves like NoEviction if there are no keys to evict with an expiration set.
+	VolatileRandom EvictionPolicy = "volatile-random"
+
+	// NoEviction does not evict any keys, and instead returns an error to the client
+	// when the max memory limit is reached.
+	NoEviction EvictionPolicy = "noeviction"
+)
+
+//publicapigen:keep
+type constStr string
+
+// Cluster represents a Redis cache cluster.
+type Cluster struct {
+	cfg ClusterConfig
+	mgr *Manager
+	cl  *redis.Client
+}
+
+// KeyspaceConfig specifies the configuration options for a cache keyspace.
+type KeyspaceConfig struct {
+	// KeyPattern is a string literal representing the
+	// cache key pattern for this keyspace.
+	KeyPattern constStr
+
+	// DefaultExpiry specifies the default key expiry for cache items
+	// in this keyspace.
+	//
+	// When set, all write operations set (for new keys)
+	// or update (for existing keys) the expiration time.
+	//
+	// When updating the expiration time Encore always
+	// performs the combined operation atomically.
+	//
+	// If nil, cache items have no expiry date by default.
+	//
+	// The default behavior can be overridden by passing in
+	// an ExpiryFunc or KeepTTL as a WriteOption to a specific operation.
+	DefaultExpiry ExpiryFunc
+
+	// EncoreInternal_DefLoc specifies where the keyspace is defined.
+	// It's an internal field set by Encore's compiler.
+	//publicapigen:drop
+	EncoreInternal_DefLoc int32
+
+	// EncoreInternal_KeyMapper specifies how typed keys are translated
+	// to a string. It's of type any to avoid making KeyspaceConfig
+	// a generic type. It's an internal field set by Encore's compiler.
+	//
+	// The type must be func(K) string.
+	//publicapigen:drop
+	EncoreInternal_KeyMapper any
+}
+
+// An OpError describes the operation that failed.
+type OpError struct {
+	Operation string
+	RawKey    string
+	Err       error
+}
+
+func (e *OpError) Error() string {
+	return fmt.Sprintf("cache: %s %q: %v", e.Operation, e.RawKey, e.Err)
+}
+
+func (e *OpError) Unwrap() error {
+	return e.Err
+}
+
+// Miss is the error value reported when a key is missing from the cache.
+// It must be checked against with errors.Is.
+var Miss = errors.New("cache miss")
+
+// KeyExists is the error reported when a key already exists
+// and the requested operation is specified to only apply to
+// keys that do not already exist.
+// It must be checked against with errors.Is.
+var KeyExists = errors.New("key already exists")
+
+// An WriteOption customizes the behavior of a single cache write operation.
+type WriteOption interface {
+	//publicapigen:keep
+	writeOption() // ensure only our package can implement
```

Comment:
I believe you can still embed this interface and calling writeOption will result in an nil pointer (or at least this is what you told me when I did this in a previous PR)

So this doesn't prevent other packages from implementing

Reply:
Yes, that's correct, Ibut it's purely meant as a signal to not create your own implementations so that's fine